### ISSUE TITLE
NodeSchool Windows Application - Looking For Feedback

### ISSUE THREAD

**Issue by andrewburgess** [nodeschool/discussions#1219]
This came about as part of a discussion during NodeConf regarding some of the difficulties with getting Windows users up to speed.

I created a weird proof of concept where Node code can be executed within the context of a .NET application. I'm looking for feedback and/or discussion about whether this is a good idea or not.

Here: https://github.com/andrewburgess/nodeschool-windows

One of the immediate benefits right now is that a user would not need to get Node installed and put on the `PATH`, nor need to open up a command line. The application uses https://github.com/tjanczuk/edge to essentially allow interop between the CLR and Node.

I have no idea where to really go with this, but maybe it can be a jumping off point to help get Windows users up and running relatively easily.

:sunglasses:

**Comment by jesstelford** [comment]
I've seen the same kind of need (as the only JS dev in a .net shop).

Unfortunately Edge was a non-starter for us: it had too many execution bugs
and didn't support iojs (pure code not producing reliable results on
consecutive executions, etc). For this reason I held off ever exposing .net
devs to Edge in case they thought of it as a reflection on node/JS in
general.

Instead, I've opted for very simple setup guides in my repos that may
require them to participate, such as:

```
git clone ...
cd ...
npm install
npm start
```

Although even this has been met with blank stares and questions such as
"oh, I can't run it in VS?", and "wow, I have to remember these commands
every time I want to run it? Can't I use a menu item..?"

A perfect example of this is [a recent tweet I made](
https://twitter.com/jesstelford/status/609198527822979074). It's also
reflected in [the way .net devs think of Javascript and it's ecosystem](
https://twitter.com/jesstelford/status/606001981379846145).

All this is to say: I agree with you; Windows devs are not going to cone to
where we are. We must go to where the Windows devs are.

I think [Node.js Tools For Visual Studio](
https://www.visualstudio.com/en-us/features/node-js-vs.aspx) is probably
the best place to start. Keep the Windows devs in an environment they're
familiar with, and slowly introduce the basic concepts to them there. It
also has the added benefit of coming with the "Made by Microsoft, therefore
it must mean it's good" stamp I've noticed .net devs appear to love.

A couple possible suggestions on how we meet them there:

1) Add Visual Studio projects to all nodeschool repos, allowing devs to
checkout and contribute

2) Platform detection for Windows boxes that includes more instruction than
`npm install -g ...`

**Comment by snbfan** [comment]
This is what I get when verifying HTTP_CLIENT exercise :

C:\projects\node_practice>learnyounode verify 9.js
_http_client.js:75
    throw new Error('Protocol "' + protocol + '" not supported. ' +
          ^
Error: Protocol "c:" not supported. Expected "http:".
    at new ClientRequest (_http_client.js:75:11)
    at Object.exports.request (http.js:49:10)
    at Object.exports.get (http.js:53:21)
    at Object.<anonymous> (C:\projects\node_practice\9.js:2:6)
    at Module._compile (module.js:460:26)
    at Object.Module._extensions..js (module.js:478:10)
    at Module.load (module.js:355:32)
    at Function.Module._load (module.js:310:12)
    at Function.Module.runMain (module.js:501:10)
    at startup (node.js:129:16)

**Comment by dominhhai** [comment]
@snbfan if you need help, please open a new issue.

**Comment by andrewburgess** [comment]
@jesstelford Totally agree with you on your points, and have experienced it myself. Part of the discussion though was focused on users who aren't even Windows devs, but are merely Windows users (i.e. not even aware of the command line). I think there were a few examples thrown around about NodeSchool events where some of the Windows users spent the majority of their time just trying to get their system set up.

Perhaps my idea might be too ambitious, but I'm hoping that we can potentially come up with some way to ease in the Windows users who are curious about software development in some capacity without throwing too many tasks in their way before they can write their first "hello world"

**Comment by jesstelford** [comment]
> focused on users who aren't even Windows devs, but are merely Windows users

Ahh, that's another kettle of fish!

To be honest, though, I haven't had too much problem helping walk complete beginners through the steps of setting up node.js and then running commands via the cmd.exe prompt.

Maybe it's more a matter of needing mentors who can dedicate some time (nodeschool workshop style) to getting them setup?

What about an online hosted workshop where users can (very easily) share their screen with the nodeschool mentor, allowing quick 1-on-1's as well as group learning? Like a MOOC for nodeschool?

But, I'm digressing... I do like your idea of trying to stick to windows only tooling. The more native the better!

I'm interested on thoughts from the others you spoke with at Nodeschool (can you @ mention them?)

**Comment by andrewburgess** [comment]
Unfortunately I can't speak more to specific issues or examples beyond what I heard. I figured I would try to help kickstart something since I develop on Windows about 40% of the time.

I also am extremely terrible with names and can't remember anyone beyond @maxogden (who I showed my proof of concept to after the fact) and @mafintosh (who was helping facilitate the discussion at the time). Maybe they remember some of the other participants from that particular talk?
